### PR TITLE
ci: migrate from CircleCI to GitHub Actions

### PR DESCRIPTION
**Pull request by Juneezee**
Completely migrates from CircleCI to GitHub Actions. Runs `lint` and `test` on every push and pull request.

### CODE REVIEW

**Review by Fontinalis — COMMENTED** [review]
I do like the idea of having everything in GitHub Actions and remove CircleCI from the list of tools we use and it's a great PR to replace tests on commits and PRs! 👍🏻 

@dhui Thoughts?

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - uses: actions/checkout@v2
+
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v2
+
+  test:
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        go: ["1.15", "1.16"]
```
Comment (dhui) [inline .github/workflows/ci.yaml:20]: Use `["1.15.x", "1.16.x"]`

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for the PR and helping us improve our release process @Juneezee !

@Fontinalis 
> I do like the idea of having everything in GitHub Actions and remove CircleCI from the list of tools we use and it's a great PR to replace tests on commits and PRs! 👍🏻
> 
> @dhui Thoughts?

Agreed, we should move our tests to use GitHub Actions as well.
Let's keep the other CI and docker files for now in case we need to continue using them. We can clean them up later.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+name: CI
```
[the remark this answers is not in the document]
Comment (Fontinalis) [reply .github/workflows/ci.yaml:1]: Personally I would prefer to keep it separate, but that could be just my taste.

[the remark this answers is not in the document]
Comment (dhui) [reply .github/workflows/ci.yaml:1]: Let's leave them separate for now then

**Review by dhui — CHANGES_REQUESTED** [review]
Add [`workflow_run`](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run) to the GoReleaser workflow/job (`.github/workflows/release.yaml`) so that it only runs after tests pass

**Comment by Juneezee** [comment]
> Add [`workflow_run`](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run) to the GoReleaser workflow/job (`.github/workflows/release.yaml`) so that it only runs after tests pass

@dhui There is a caveat if we use [`workflow_run`](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run) to trigger the `goreleaser` job, as described in [this StackOverflow answer](https://stackoverflow.com/a/68078768/7902371).

If the `goreleaser` job is trigerred by `workflow_run` instead of `on: push: tags: "v*"`, GoReleaser would not be able to detect tag that is pushed, which would then cause the job to fail.

**Comment by Fontinalis** [comment]
> If the goreleaser job is trigerred by workflow_run instead of on: push: tags: "v*", GoReleaser would not be able to detect tag that is pushed, which would then cause the job to fail.

On the other hand, this could be solved if we merge the two workflows into one, and use the following code by @Juneezee  to check if the release should run or not 

```diff
  goreleaser:
    name: Release a new version
    environment: GoReleaser
    runs-on: ubuntu-latest
+   needs: [lint, test]
+   if: ${{ success() && github.repository == 'golang-migrate/migrate' && contains(github.ref, 'refs/tags/v') }}
    ...
```

So I'm happy to turn my vote on the question of "should we merge the two workflows or not".

**Review by Fontinalis — APPROVED** [review]
Looks legit to me! @dhui you concur?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -1,13 +1,58 @@
...
+          path-to-profile: /tmp/coverage/combined.txt
+          flag-name: Go-${{ matrix.go }}
+          parallel: true
+
+  test-finish:
+    needs: [test]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: shogo82148/actions-goveralls@v1
+        with:
```
Comment (Juneezee) [inline .github/workflows/ci.yaml:43]: I missed these lines in the previous commit. They are required for coveralls to work correctly with parallel tests, see https://github.com/shogo82148/actions-goveralls#parallel-job-example

**Comment by Juneezee** [comment]
Last round of review I think, everything seems to be well configured :rocket:

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -1,13 +1,58 @@
...
+        run: make test COVERAGE_DIR=/tmp/coverage
+
+      - name: Send goveralls coverage
+        uses: shogo82148/actions-goveralls@v1
+        with:
+          path-to-profile: /tmp/coverage/combined.txt
+          flag-name: Go-${{ matrix.go }}
+          parallel: true
+
+  test-finish:
```
Comment (dhui) [inline .github/workflows/ci.yaml:38]: ~Rename to `send-coverage` and add a human-readable `name: Send coverage`~

EDIT: This looks like a duplicate sending of coverage?

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -1,13 +1,58 @@
...
     name: Release a new version
-    environment: GoReleaser
+    needs: [lint, test]
     runs-on: ubuntu-latest
+    environment: GoReleaser
+    # This job only runs when
+    # 1. When the previous `lint` and `test` jobs has completed successfully
+    # 2. When the repository is not a fork, i.e. it will only run on the official golang-migrate/migrate
+    # 3. When the workflow is triggered by a tag pushed with `v` prefix
+    if: ${{ success() && github.repository == 'golang-migrate/migrate' && contains(github.ref, 'refs/tags/v') }}
```
Comment (dhui) [inline .github/workflows/ci.yaml:55]: Use `startsWith()` instead of `contains()`

**Review by dhui — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (dhui) [reply .github/workflows/ci.yaml:38]: Oh, NVM, it looks like this notifies github that coverage has pass. Let's name this `check-coverage` with a human readable `name: Check coverage`

**Comment by Juneezee** [comment]
@dhui Please take a look again :smiley:

**Review by dhui — APPROVED** [review]
Thanks for migrating our tests over!

**Pull request merged**